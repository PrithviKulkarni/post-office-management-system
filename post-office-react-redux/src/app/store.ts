import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import letterReducer from '../features/LetterSlice';
import parcelReducer from '../features/ParcelSlice';
import depotReducer from '../features/DepotSlice';
import driverReducer from '../features/DriverSlice' 
export const store = configureStore({
  reducer: {
    letter: letterReducer,
    parcel: parcelReducer,
    depot: depotReducer,
    driver: driverReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
