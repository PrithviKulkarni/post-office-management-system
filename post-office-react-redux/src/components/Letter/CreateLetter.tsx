import { useNavigate } from "react-router-dom";
import Button from "react-bootstrap/Button";
import { useState } from "react";
import Form from "react-bootstrap/Form";
import { useAppDispatch } from "../../app/hooks";
import { addNewLetter } from "../../features/LetterSlice";
import { defaultLetter, Letter } from "../interfaces/letterInterface";

/**
 *
 * @returns A form to fill by the user to create a new chat
 */
const CreateLetter = () => {
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const [address, setAddress] = useState<string>("");
  const [deliveryClass, setDeliveryClass] = useState<string>("");
  const [addressBool, setAddressBool] = useState<boolean>(false);
  const [deliveryClassBool, setDeliveryClassBool] = useState<boolean>(false);

  const handleSubmit = (event: any) => {
    event.preventDefault();

    const validated = [address, deliveryClass].every(Boolean);

    if (!address) {
      setAddressBool(true);
    } else {
      setAddressBool(false);
    }
    if (!deliveryClass) {
      setDeliveryClassBool(true);
    } else {
      setDeliveryClassBool(false);
    }

    if (validated) {
      const updLetter: Letter = defaultLetter;
      if (address) {
        updLetter.address = address;
      }
      if (deliveryClass) {
        updLetter.delivery_class = deliveryClass;
      }

      dispatch(addNewLetter(updLetter));
      navigate("/letters");
    }
  };

  return (
    <div className="createLetter" data-testid="createLetter">
      <Form
        onSubmit={handleSubmit}
        style={{
          height: "100%",
          display: "flex",
          flexDirection: "column",
          justifyContent: "center",
          alignItems: "center",
          width: "100%",
        }}
      >
        <Form.Group className="mb-3" controlId="formMessageFrom">
          <Form.Label>Address</Form.Label>
          <Form.Control
            type="text"
            data-testid="address-input"
            value={address}
            onChange={(event) => setAddress(event.target.value)}
          />
          {addressBool && (
            <Form.Text className="text-muted" data-testid="validate-address">Please fill in field</Form.Text>
          )}
        </Form.Group>

        <Form.Group className="mb-3" controlId="formMessageContent">
          <Form.Label>Delivery Class</Form.Label>
          <Form.Control
            type="text"
            data-testid="delivery-input"
            value={deliveryClass}
            onChange={(event) => setDeliveryClass(event.target.value)}
          />
          {deliveryClassBool && (
            <Form.Text className="text-muted" data-testid="validate-delivery">Please fill in field</Form.Text>
          )}
        </Form.Group>
        <Button variant="primary" type="submit" className="createButton" data-testid="createButton">
          Submit
        </Button>
      </Form>
    </div>
  );
};

export default CreateLetter;
