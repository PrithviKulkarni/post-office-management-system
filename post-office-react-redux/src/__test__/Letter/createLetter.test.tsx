import React, { useState } from "react";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { store } from "../../app/store";
import { Provider } from "react-redux";
import CreateLetter from "../../components/Letter/CreateLetter";
import { BrowserRouter } from "react-router-dom";

 afterEach(cleanup);
it("renders Create Letter page correctly", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <BrowserRouter>
        <CreateLetter />
      </BrowserRouter>
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("Verifies the Create Letter render", () => {
  const { getByTestId } = render(
    <Provider store={store}>
      <BrowserRouter>
        <CreateLetter />
      </BrowserRouter>
    </Provider>
  );
  const button = getByTestId("createButton");
  expect(button.textContent).toBe("Submit");
  fireEvent.click(button);
  expect(button).toBeTruthy();
  const address = getByTestId("validate-address");
  const delivery = getByTestId("validate-delivery");
  expect(address.textContent).toBe("Please fill in field");
  expect(delivery.textContent).toBe("Please fill in field");
  expect(screen.getByTestId("createLetter")).toHaveClass("createLetter");
});

it("Verifies the Filled Create Letter render", () => {
  const { getByTestId } = render(
    <Provider store={store}>
      <BrowserRouter>
        <CreateLetter />
      </BrowserRouter>
    </Provider>
  );
  const button = getByTestId("createButton");
  expect(button.textContent).toBe("Submit");
  const address = getByTestId("address-input");
  const delivery = getByTestId("delivery-input");
  fireEvent.change(address, { target: { value: "test" } });
  fireEvent.change(delivery, { target: { value: "test" } });
  fireEvent.click(button);
  expect(screen.getByTestId("createLetter")).toHaveClass("createLetter");
});
