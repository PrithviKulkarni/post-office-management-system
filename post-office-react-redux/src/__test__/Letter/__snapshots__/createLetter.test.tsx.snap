// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Create Letter page correctly 1`] = `
<DocumentFragment>
  <div
    class="createLetter"
    data-testid="createLetter"
  >
    <form
      class=""
      style="height: 100%; display: flex; flex-direction: column; justify-content: center; align-items: center; width: 100%;"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formMessageFrom"
        >
          Address
        </label>
        <input
          class="form-control"
          data-testid="address-input"
          id="formMessageFrom"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formMessageContent"
        >
          Delivery Class
        </label>
        <input
          class="form-control"
          data-testid="delivery-input"
          id="formMessageContent"
          type="text"
          value=""
        />
      </div>
      <button
        class="createButton btn btn-primary"
        data-testid="createButton"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
